### DISCUSSION TITLE
Is it possible to preemptively provide file data?

### DISCUSSION THREAD

**Discussion by ZabirHoque** [microsoft/ProjFS-Managed-API#84] [Q&A] [answered]
ATM, I'm preemptively populating files by running a separate daemon process that touches desired files. I'm wondering if there is a more efficient way to do this?

`PrjWriteFileData()` seems to be the only way to provide file data to the underlying system, but its not clear to me how to derive the `dataStreamId` parameter when not responding to a `PRJ_GET_FILE_DATA_CB` callback.

**Comment by cgallred** [comment, marked as answer]
Placeholders can be preemptively laid down with `PrjWritePlaceholderInfo`/`PrjWritePlaceholderInfo2`, but files can't be preemptively hydrated in the same way. When a user tries to read a non-hydrated placeholder, the ProjFS file system filter driver generates a `dataStreamId` value, associates it with the user's FILE_OBJECT, and sends the `PRJ_GET_FILE_DATA_CB` callback. The hydration mechanism uses that FILE_OBJECT to write the data. There isn't currently a way to get such a FILE_OBJECT outside the regular hydration path that your daemon triggers.
